chromadb/segment/impl/vector/brute_force_index.py

```diff
@@ -0,0 +1,151 @@
+from typing import Any, Callable, Dict, List, Optional, Sequence, Set
+import numpy as np
+import numpy.typing as npt
+from chromadb.types import (
+    EmbeddingRecord,
+    VectorEmbeddingRecord,
+    VectorQuery,
+    VectorQueryResult,
+)
+
+from chromadb.utils import distance_functions
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class BruteForceIndex:
+    """A lightweight, numpy based brute force index that is used for batches that have not been indexed into hnsw yet"""
+
+    id_to_index: Dict[str, int]
+    index_to_id: Dict[int, str]
+    id_to_seq_id: Dict[str, int]
+    deleted_ids: Set[str]
+    free_indices: List[int]
+    size: int
+    dimensionality: int
+    distance_fn: Callable[[npt.NDArray[Any], npt.NDArray[Any]], float]
+    vectors: npt.NDArray[Any]
+
+    def __init__(self, size: int, dimensionality: int, space: str = "l2"):
+        if space == "l2":
+            self.distance_fn = distance_functions.l2
+        elif space == "ip":
+            self.distance_fn = distance_functions.ip
+        elif space == "cosine":
+            self.distance_fn = distance_functions.cosine
+        else:
+            raise Exception(f"Unknown distance function: {space}")
+
+        self.id_to_index = {}
+        self.index_to_id = {}
+        self.id_to_seq_id = {}
+        self.deleted_ids = set()
+        self.free_indices = list(range(size))
+        self.size = size
+        self.dimensionality = dimensionality
+        self.vectors = np.zeros((size, dimensionality))
+
+    def __len__(self) -> int:
+        return len(self.id_to_index)
+
+    def flush(self) -> None:
+        self.id_to_index = {}
+        self.index_to_id = {}
+        self.id_to_seq_id = {}
+        self.deleted_ids.clear()
+        self.free_indices = list(range(self.size))
+        self.vectors.fill(0)
+
+    # TODO: thread safety
+    def upsert(self, records: List[EmbeddingRecord]) -> None:
+        if len(records) + len(self) > self.size:
+            raise Exception(
+                "Index with capacity {} and {} current entries cannot add {} records".format(
+                    self.size, len(self), len(records)
+                )
+            )
+
+        for i, record in enumerate(records):
+            id = record["id"]
+            vector = record["embedding"]
+            self.id_to_seq_id[id] = record["seq_id"]
+            if id in self.deleted_ids:
+                self.deleted_ids.remove(id)
+
+            if id in self.id_to_index:
```

Comment:
Add a comment that we may want to update numpy using multi-index selection which may be faster.